File: packages/javascript/examples/grain-marks/grain-marks.js
import { seededTrianglePoints2D, mapTriangleCoordinates2D } from "../../src/triangle-points.js";
import { seededQuadrantPartition2D } from "../../src/quadrant-partition.js";

/** Editable grain composition motivated by puntis and puntis3.
 * Density, triangle construction and biased expressions are example choices.
 * Sampling/mapping operations retain points; content and style remain independent.
 */
export function createGrainComposition(seed, density, distribution, cells) {
  if (!Number.isInteger(seed) || seed < 0 || seed > 0xffffffff
      || !Number.isFinite(density) || density < 0
      || !Number.isInteger(distribution) || distribution < 0 || distribution > 2
      || typeof cells !== "boolean") throw new Error("Invalid example configuration");
  const triangles = [];
  if (!cells) triangles.push([[40, 600], [320, 40], [600, 600]]);
  else {
    const layout = seededQuadrantPartition2D({seed, replacements: 4, selectionFraction: 0.5,
      origin: [0, 0], extent: [640, 640]});
    const bounds = new Float64Array(4);
    for (let i = 0; i < layout.size; i++) {
      layout.boundsInto(i, bounds);
      const [l, t, r, b] = bounds;
      triangles.push([[l, t], [r, t], [r, b]], [[l, t], [r, b], [l, b]]);
    }
  }
  let totalPoints = 0;
  const counts = triangles.map(([a, b, c]) => {
    const area = Math.abs((b[0] - a[0]) * (c[1] - a[1])
      - (b[1] - a[1]) * (c[0] - a[0])) * 0.5;
    const count = Math.ceil(area * density);
    if (!Number.isFinite(count) || count > 160000 - totalPoints)
      throw new Error("Example exceeds 160000-point work budget");
    totalPoints += count;
    return count;
  });
  const regions = triangles.map((triangle, index) => {
    const regionSeed = (seed + index) >>> 0;
    if (distribution === 0)
      return seededTrianglePoints2D({seed: regionSeed, count: counts[index], triangle});
    const random = javaExampleRandom(regionSeed);
    const unitCoordinates = [];
    for (let p = 0; p < counts[index]; p++) {
      let u, v;
      if (distribution === 1) {
        u = random() * random();
        v = random();
      } else {
        const side = random() < 0.5 ? 0 : 1;
        const lower = side * 0.8;
        v = (lower + (1 - lower) * random()) * (0.4 + 0.6 * random());
        u = random();
      }
      unitCoordinates.push([u, v]);
    }
    return mapTriangleCoordinates2D({triangle, unitCoordinates});
  });
  return Object.freeze({size: regions.length, totalPoints,
    regionAt(index) { return regions[index]; }});
}

// Example-only java.util.Random nextDouble sequence, matching GrainComposition.
// This does not expose or alter the library's private xoshiro sampling stream.
function javaExampleRandom(seed) {
  const multiplier = 0x5deece66dn, mask = (1n << 48n) - 1n;
  let state = (BigInt(seed) ^ multiplier) & mask;
  function next(bits) {
    state = (state * multiplier + 11n) & mask;
    return Number(state >> BigInt(48 - bits));
  }
  return () => (next(26) * 134217728 + next(27)) / 9007199254740992;
}

File: packages/python/examples/grain_marks/grain_marks.py
"""Example-owned retained grain composition matching Java GrainComposition.

The canvas, density, and two biased coordinate expressions are choices for the
editable GrainMarks example.  They are not sampling-operation defaults.
"""
from dataclasses import dataclass
import math

from procedurals.quadrant_partition import seeded_quadrant_partition_2d
from procedurals.triangle_points import (
    map_triangle_coordinates_2d,
    seeded_triangle_points_2d,
)

_MASK48 = (1 << 48) - 1
_MULTIPLIER = 0x5DEECE66D
_POINT_BUDGET = 160_000


class _JavaExampleRandom:
    """The example's private java.util.Random-compatible nextDouble stream."""
    def __init__(self, seed):
        self._state = (seed ^ _MULTIPLIER) & _MASK48

    def _next(self, bits):
        self._state = (self._state * _MULTIPLIER + 11) & _MASK48
        return self._state >> (48 - bits)

    def next_double(self):
        return (self._next(26) * 134217728 + self._next(27)) / 9007199254740992.0


@dataclass(frozen=True, slots=True)
class _GrainComposition:
    _regions: tuple
    _total_points: int

    @property
    def size(self):
        return len(self._regions)

    @property
    def total_points(self):
        return self._total_points

    def region_at(self, index):
        return self._regions[index]


def _triangle(ax, ay, bx, by, cx, cy):
    return [[ax, ay], [bx, by], [cx, cy]]


def _triangles(seed, cells):
    if not cells:
        return [_triangle(40.0, 600.0, 320.0, 40.0, 600.0, 600.0)]
    layout = seeded_quadrant_partition_2d({
        "seed": seed, "replacements": 4, "selectionFraction": 0.5,
        "origin": [0.0, 0.0], "extent": [640.0, 640.0],
    })
    triangles, bounds = [], [0.0] * 4
    for index in range(layout.size):
        layout.bounds_into(index, bounds)
        left, top, right, bottom = bounds
        triangles.append(_triangle(left, top, right, top, right, bottom))
        triangles.append(_triangle(left, top, right, bottom, left, bottom))
    return triangles


def _counts(triangles, density):
    counts, total = [], 0
    for triangle in triangles:
        a, b, c = triangle
        area = abs((b[0] - a[0]) * (c[1] - a[1]) - (b[1] - a[1]) * (c[0] - a[0])) * 0.5
        requested = area * density
        if not math.isfinite(requested):
            raise ValueError("example exceeds 160000-point work budget")
        count = math.ceil(requested)
        if count > _POINT_BUDGET - total:
            raise ValueError("example exceeds 160000-point work budget")
        counts.append(count)
        total += count
    return counts, total


def _biased_units(seed, count, distribution):
    random = _JavaExampleRandom(seed)
    units = []
    for _ in range(count):
        if distribution == 1:
            u = random.next_double() * random.next_double()
            v = random.next_double()
        else:
            side = 0 if random.next_double() < 0.5 else 1
            lower = side * 0.8
            v = (lower + (1.0 - lower) * random.next_double()) * (0.4 + 0.6 * random.next_double())
            u = random.next_double()
        units.append([u, v])
    return units


def create_grain_composition(seed, density, distribution, cells):
    """Build retained grain; motivated by survey/out/2018/Generativos/puntis/notes.md
    and puntis3/notes.md. Density is an example choice, not a measured useful range.

    Distribution zero uses the seeded public core.  One and two materialize the
    example's Java-Random caller pairs then use the explicit mapping core.
    """
    if type(density) in (int, float):
        try:
            density_value = float(density)
        except OverflowError:
            density_value = math.inf
    else:
        density_value = math.nan
    if (type(seed) is not int or seed < 0 or seed > 0xFFFFFFFF
            or not math.isfinite(density_value) or density_value < 0
            or type(distribution) is not int or distribution < 0 or distribution > 2
            or type(cells) is not bool):
        raise ValueError("invalid example configuration")
    triangles = _triangles(seed, cells)
    counts, total = _counts(triangles, density_value)
    regions = []
    for index, triangle in enumerate(triangles):
        region_seed = (seed + index) & 0xFFFFFFFF
        if distribution == 0:
            regions.append(seeded_triangle_points_2d({
                "seed": region_seed, "count": counts[index], "triangle": triangle,
            }))
        else:
            regions.append(map_triangle_coordinates_2d({
                "triangle": triangle,
                "unitCoordinates": _biased_units(region_seed, counts[index], distribution),
            }))
    return _GrainComposition(tuple(regions), total)
